spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/InferShape.scala

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.nn.abstractnn
+
+import com.intel.analytics.bigdl.utils.Shape
+
+trait InferShape {
+
+  private var _inputShapeValue: Shape = null
+
+  private var _outputShapeValue: Array[Shape] = Array[Shape]()
+
+  private[bigdl] def inputShapeValue: Shape = _inputShapeValue
+
+  private[bigdl] def outputShapeValue: Array[Shape] = _outputShapeValue
+
+  // scalastyle:off
+  private[bigdl] def inputShapeValue_=(value: Shape): Unit = {
+    _inputShapeValue = value
+  }
+
+  private[bigdl] def outputShapeValue_=(value: Array[Shape]): Unit = {
+    _outputShapeValue = value
+  }
+  // scalastyle:on
+
+  /**
+   * We suppose the first dim is batch
+   */
+  private[bigdl] def getInputShape(): Shape = {
+    _inputShapeValue
+  }
+
+  /**
+   * Get the outputshape by index.
+   * @param index start from 0
+   * @return
+   */
+  private[bigdl] def getOutputShapeFor(index: Int): Shape = {
+    _outputShapeValue(index)
+  }
+
+  /**
+   * We suppose the first dim is batch
+   */
+  private[bigdl] def getOutputShape(): Shape = {
+    if (_outputShapeValue.length > 1) {
+      throw new RuntimeException(
+        "There's multipule output for this layer. Please use getInputShapeFor instead")
+    }
+    outputShapeValue(0)
+  }
+
+  /**
+   * Execute builing logic and return the outputShape for the given inputShape.
+   * NB: the first dim of inputShape is batch
+   */
+  private[bigdl] def build(inputShape: Shape): Shape = {
```

Comment:
I don't think we should call build multiple times; check if `isBuilt` is true here